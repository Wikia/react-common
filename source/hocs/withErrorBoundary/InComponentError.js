import React from 'react';
import styled from 'styled-components';

const InComponentWrapper = styled.div`
    background: #fff;
    padding: 10px;
`;

const Heading = styled.h4`
    align-content: center;
    color: red;
    display: flex;
    margin-top: 0;
    margin-bottom: 0;
    padding: 10px;
`;

export default function DefaultFallbackComponent() {
    return (
        <InComponentWrapper>
            <Heading>
                Sorry, weâ€™re having trouble displaying this content right now
            </Heading>
        </InComponentWrapper>
    );
}
