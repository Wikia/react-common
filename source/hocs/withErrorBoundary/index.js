import PropTypes from 'prop-types';
import React from 'react';

import logErrorEvent from '../../utils/eventLogger';

import InComponentErrorBoundary from './InComponentError';

class ErrorBoundary extends React.Component {
    static getDerivedStateFromError(error) {
        return { hasError: true, error };
    }

    constructor(props) {
        super(props);
        this.state = {
            hasError: false,
            error: null,
            info: null,
        };
    }

    componentDidCatch(error, info) {
        console.log(`Error Boundary (${this.props.boundaryName}):`, error, info);
        this.setState({ info });
        logErrorEvent(this.props.appName, `error-boundary-${this.props.boundaryName}`, {
            errorStack: error.stack,
            componentStack: info.componentStack,
        });
    }

    render() {
        const FallbackComponent = this.props.fallbackComponent;

        if (this.state.hasError) {
            if (!FallbackComponent) {
                return false;
            }

            return <FallbackComponent error={this.state.error} info={this.state.info} />;
        }

        return this.props.children;
    }
}

ErrorBoundary.propTypes = {
    appName: PropTypes.string.isRequired,
    boundaryName: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
    fallbackComponent: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.func,
    ]).isRequired,
};

export default function withErrorBoundary(WrappedComponent, boundaryName, appName, FallBackComponent = InComponentErrorBoundary) {
    const Component = props => (
        <ErrorBoundary fallbackComponent={FallBackComponent} boundaryName={boundaryName} appName={appName}>
            <WrappedComponent {...props} />
        </ErrorBoundary>
    );

    return Component;
}
