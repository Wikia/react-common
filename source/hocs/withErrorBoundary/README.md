Add an error boundary to any component

## Usage

```js static
import withErrorBoundary from '@react-common/hocs/withErrorBoundary';

const Component = props => (<div> Test </div>);
const ComponentWithErrorBoundary = withErrorBoundary(Component, 'RequiredBoundaryName');

// Usage
<ComponentWithErrorBoundary />
```

Custom Fallback 

```js static
import withErrorBoundary from '@react-common/hocs/withErrorBoundary';

const Component = props => (<div> Test </div>);
const AlternateFallBack = props => (<div> Alternate error message </div>);
const ComponentWithErrorBoundary = withErrorBoundary(Component, 'RequiredBoundaryName', AlternateFallback);

// Usage
<ComponentWithErrorBoundary />
```
