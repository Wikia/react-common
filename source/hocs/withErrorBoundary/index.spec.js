import { mount } from 'enzyme/build/index';
import React from 'react';

import withErrorBoundary from './index';

function Something() {
    // this is just a placeholder
    throw new Error('test');
}
const SimpleComponent = () => (
    <div>
        <Something />
    </div>
);

test('Default fallback error boundary', () => {
    const ComponentWithErrorBoundary = withErrorBoundary(SimpleComponent, 'TestBoundaryName', 'FandomCreator', 'v1');

    const component = mount(
        <ComponentWithErrorBoundary />,
    );


    expect(component).toMatchSnapshot();
});

test('Renders nothing when there is undefined fallback', () => {
    const ComponentWithErrorBoundary = withErrorBoundary(SimpleComponent, 'TestBoundaryName', 'FEPO', 'v2', null);

    const component = mount(
        <ComponentWithErrorBoundary />,
    );

    expect(component).toMatchSnapshot();
});
