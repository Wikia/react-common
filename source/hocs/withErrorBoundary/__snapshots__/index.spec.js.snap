// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default fallback error boundary 1`] = `
<Component>
  <ErrorBoundary
    appName="FandomCreator"
    boundaryName="TestBoundaryName"
    fallbackComponent={[Function]}
  >
    <DefaultFallbackComponent
      error={[Error: test]}
      info={
        Object {
          "componentStack": "
    in Something (created by SimpleComponent)
    in div (created by SimpleComponent)
    in SimpleComponent (created by Component)
    in ErrorBoundary (created by Component)
    in Component (created by WrapperComponent)
    in WrapperComponent",
        }
      }
    >
      <styled.div>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": true,
                "lastClassName": "gQXkGF",
                "rules": Array [
                  "
    background: #fff;
    padding: 10px;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="sc-bdVaJa gQXkGF"
          >
            <styled.h4>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": true,
                      "lastClassName": "ePWlJX",
                      "rules": Array [
                        "
    align-content: center;
    color: red;
    display: flex;
    margin-top: 0;
    margin-bottom: 0;
    padding: 10px;
",
                      ],
                    },
                    "displayName": "styled.h4",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": "h4",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <h4
                  className="sc-bwzfXH ePWlJX"
                >
                  Sorry, weâ€™re having trouble displaying this content right now
                </h4>
              </StyledComponent>
            </styled.h4>
          </div>
        </StyledComponent>
      </styled.div>
    </DefaultFallbackComponent>
  </ErrorBoundary>
</Component>
`;

exports[`Renders nothing when there is undefined fallback 1`] = `
<Component>
  <ErrorBoundary
    appName="FEPO"
    boundaryName="TestBoundaryName"
    fallbackComponent={null}
  />
</Component>
`;
